File: backend/tests/test_export.py
"""Tests for GET /api/export/{item_id:path}."""
import pytest
from datetime import datetime


def test_export_schema_version(client, plan_id):
    r = client.get(f"/api/export/{plan_id}")
    assert r.status_code == 200
    data = r.json()
    assert data["schemaVersion"] == "1.0"


def test_export_summary_total_correct(client, plan_id):
    # Add 3 comments
    for i in range(3):
        client.post(f"/api/comments/{plan_id}", json={
            "author": f"User{i}",
            "content": f"Comment {i}",
            "reference": {"type": "general", "value": None}
        })
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert data["summary"]["total"] == 3


def test_export_summary_open_correct(client, plan_id):
    # Add 2 comments, resolve 1
    r1 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Alice",
        "content": "Open",
        "reference": {"type": "general", "value": None}
    })
    r2 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Bob",
        "content": "To resolve",
        "reference": {"type": "general", "value": None}
    })
    client.patch(f"/api/comments/{plan_id}/{r2.json()['id']}/resolve")
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert data["summary"]["open"] == 1


def test_export_summary_resolved_correct(client, plan_id):
    r1 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Carol",
        "content": "Open",
        "reference": {"type": "general", "value": None}
    })
    r2 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Dave",
        "content": "Resolved",
        "reference": {"type": "general", "value": None}
    })
    client.patch(f"/api/comments/{plan_id}/{r2.json()['id']}/resolve")
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert data["summary"]["resolved"] == 1


def test_export_open_comments_only_open_status(client, plan_id):
    r1 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Eve",
        "content": "Open",
        "reference": {"type": "general", "value": None}
    })
    r2 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Frank",
        "content": "Resolved",
        "reference": {"type": "general", "value": None}
    })
    client.patch(f"/api/comments/{plan_id}/{r2.json()['id']}/resolve")
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    open_ids = [c["id"] for c in data["openComments"]]
    assert r1.json()["id"] in open_ids
    assert r2.json()["id"] not in open_ids


def test_export_has_exported_at(client, plan_id):
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert "exportedAt" in data
    assert data["exportedAt"] is not None
    # Should be a valid datetime string
    parsed = datetime.fromisoformat(data["exportedAt"].replace("Z", "+00:00"))
    assert parsed is not None


def test_export_all_resolved_open_comments_empty(client, plan_id):
    r1 = client.post(f"/api/comments/{plan_id}", json={
        "author": "Grace",
        "content": "To resolve",
        "reference": {"type": "general", "value": None}
    })
    cid = r1.json()["id"]
    client.patch(f"/api/comments/{plan_id}/{cid}/resolve")
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert data["openComments"] == []


def test_export_nonexistent_item_returns_404(client):
    r = client.get("/api/export/nonexistent/item")
    assert r.status_code == 404


def test_export_open_comment_fields(client, plan_id):
    client.post(f"/api/comments/{plan_id}", json={
        "author": "Hank",
        "content": "Check fields",
        "reference": {"type": "section", "value": "## Overview"}
    })
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert len(data["openComments"]) == 1
    comment = data["openComments"][0]
    assert "id" in comment
    assert "author" in comment
    assert "content" in comment
    assert "createdAt" in comment
    assert comment["author"] == "Hank"
    assert comment["content"] == "Check fields"


def test_export_empty_initially(client, plan_id):
    r = client.get(f"/api/export/{plan_id}")
    data = r.json()
    assert data["summary"]["total"] == 0
    assert data["summary"]["open"] == 0
    assert data["summary"]["resolved"] == 0
    assert data["openComments"] == []

File: server/crates/revi-server/tests/test_export.rs
mod common;
use common::{body_json, AppFixture};

#[tokio::test]
async fn test_export_schema_version() {
    let fx = AppFixture::new();
    let resp = fx.get("/api/export/plans/sprint-1-design").await;
    assert_eq!(resp.status(), 200);
    let body = body_json(resp).await;
    assert_eq!(body["schemaVersion"], "1.0");
}

#[tokio::test]
async fn test_export_item_fields() {
    let fx = AppFixture::new();
    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let item = &body["item"];
    assert_eq!(item["id"], "plans/sprint-1-design");
    assert_eq!(item["type"], "plan");
    assert!(item["title"].is_string());
}

#[tokio::test]
async fn test_export_summary_zero_initially() {
    let fx = AppFixture::new();
    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let summary = &body["summary"];
    assert_eq!(summary["total"], 0);
    assert_eq!(summary["open"], 0);
    assert_eq!(summary["resolved"], 0);
}

#[tokio::test]
async fn test_export_summary_counts_correctly() {
    let fx = AppFixture::new();

    // Add 3 comments
    for i in 0..3 {
        let resp = fx
            .post_json(
                "/api/comments/plans/sprint-1-design",
                serde_json::json!({"author": "Agent", "content": format!("Comment {}", i)}),
            )
            .await;
        let c = body_json(resp).await;
        // Resolve first 2
        if i < 2 {
            let id = c["id"].as_str().unwrap();
            fx.patch_empty(&format!(
                "/api/comments/plans/sprint-1-design/{}/resolve",
                id
            ))
            .await;
        }
    }

    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let summary = &body["summary"];
    assert_eq!(summary["total"], 3);
    assert_eq!(summary["open"], 1);
    assert_eq!(summary["resolved"], 2);
}

#[tokio::test]
async fn test_export_open_comments_only() {
    let fx = AppFixture::new();

    let resp = fx
        .post_json(
            "/api/comments/plans/sprint-1-design",
            serde_json::json!({"author": "Alice", "content": "Keep this open"}),
        )
        .await;
    let c1 = body_json(resp).await;

    let resp = fx
        .post_json(
            "/api/comments/plans/sprint-1-design",
            serde_json::json!({"author": "Bob", "content": "Close this"}),
        )
        .await;
    let c2 = body_json(resp).await;
    let c2_id = c2["id"].as_str().unwrap();
    fx.patch_empty(&format!(
        "/api/comments/plans/sprint-1-design/{}/resolve",
        c2_id
    ))
    .await;

    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let open = body["openComments"].as_array().unwrap();
    assert_eq!(open.len(), 1);
    assert_eq!(open[0]["id"], c1["id"]);
}

#[tokio::test]
async fn test_export_open_comments_fields() {
    let fx = AppFixture::new();
    fx.post_json(
        "/api/comments/plans/sprint-1-design",
        serde_json::json!({
            "author": "Agent",
            "content": "Needs work",
            "reference": {"type": "section", "value": "## Overview"}
        }),
    )
    .await;

    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let comment = &body["openComments"][0];
    assert!(comment.get("id").is_some());
    assert_eq!(comment["author"], "Agent");
    assert_eq!(comment["content"], "Needs work");
    assert!(comment["createdAt"].is_string());
    assert_eq!(comment["reference"]["type"], "section");
    // Exported comments should NOT have status/resolvedAt/itemId
    assert!(comment.get("status").is_none());
    assert!(comment.get("itemId").is_none());
}

#[tokio::test]
async fn test_export_not_found() {
    let fx = AppFixture::new();
    let resp = fx.get("/api/export/plans/no-such-item").await;
    assert_eq!(resp.status(), 404);
}

#[tokio::test]
async fn test_export_has_exported_at() {
    let fx = AppFixture::new();
    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    assert!(body["exportedAt"].is_string());
}

#[tokio::test]
async fn test_export_archived_not_in_counts() {
    let fx = AppFixture::new();

    // Add and resolve a comment, then archive it
    let resp = fx
        .post_json(
            "/api/comments/plans/sprint-1-design",
            serde_json::json!({"author": "Alice", "content": "Archived"}),
        )
        .await;
    let c = body_json(resp).await;
    let id = c["id"].as_str().unwrap();
    fx.patch_empty(&format!(
        "/api/comments/plans/sprint-1-design/{}/resolve",
        id
    ))
    .await;
    fx.post_empty("/api/archive/plans/sprint-1-design").await;

    // After archiving, counts should be 0
    let body = body_json(fx.get("/api/export/plans/sprint-1-design").await).await;
    let summary = &body["summary"];
    assert_eq!(summary["total"], 0);
    assert_eq!(summary["open"], 0);
    assert_eq!(summary["resolved"], 0);
}
